src/UglyToad.PdfPig/Writer/PdfMerger.cs

```diff
@@ -286,79 +266,111 @@ private ObjectToken CopyPageNode(PageTreeNode pageNode, IndirectReferenceToken p
+                            var name = setPair.Key;
+                            var token = setPair.Value;
+                            newContent.Add(NameToken.Create(name), CopyToken(token, tokenScanner, deepCopy));
+                        }
+
+                        return new DictionaryToken(newContent);
                     }
+                    case ArrayToken arrayToken:
+                    {
+                        var newArray = new List<IToken>(arrayToken.Length);
+                        foreach (var token in arrayToken.Data)
+                        {
+                            newArray.Add(CopyToken(token, tokenScanner, deepCopy));
+                        }
 
-                    return new DictionaryToken(newContent);
-                }
-                else if (tokenToCopy is ArrayToken arrayToken)
-                {
-                    var newArray = new List<IToken>(arrayToken.Length);
-                    foreach (var token in arrayToken.Data)
+                        return new ArrayToken(newArray);
+                    }
+                    case IndirectReferenceToken referenceToken:
                     {
-                        newArray.Add(CopyToken(token, tokenScanner));
+                        if (referencesFromDocument.TryGetValue(referenceToken, out var newReferenceToken))
+                        {
+                            return newReferenceToken;
+                        }
+                        
+                        var tokenObject = DirectObjectFinder.Get<IToken>(referenceToken.Data, tokenScanner);
+
+                        Debug.Assert(!(tokenObject is IndirectReferenceToken));
+
+                        var newToken = CopyToken(tokenObject, tokenScanner, deepCopy);
+                        newReferenceToken = context.WriteToken(newToken);
+                        
+                        referencesFromDocument.Add(referenceToken, newReferenceToken);
+                        
+                        return newReferenceToken;
                     }
+                    case StreamToken streamToken:
+                    {
+                        var properties = CopyToken(streamToken.StreamDictionary, tokenScanner, deepCopy) as DictionaryToken;
+                        Debug.Assert(properties != null);
 
-                    return new ArrayToken(newArray);
+                        var bytes = deepCopy ? new List<byte>(streamToken.Data) : streamToken.Data;
+                        return new StreamToken(properties, bytes);
+                    }
+                    
+                    case ObjectToken _:
+                    {
+                        // This is because, since we don't write token directly to the stream. So can't know the offset.
+                        // The token would be invalid. Although I don't think the copy of an object token would ever happen
+                        throw new NotSupportedException("Copying a Object token is not supported");
```

Comment:
Is this clear enough?

Reply:
I think this makes sense, as you say I don't think we'd ever be copying an object token because we're creating object tokens by copying and wrapping the non-object token types.